### ISSUE TITLE
dockerd seems not to close file descriptors to /var/run/docker.sock and hits "too many open files" error

### ISSUE THREAD

**Issue by pstrzelczak** [moby/moby#37061]
**Description**
I have a docker swarm on 3 manager nodes. I am deploying a few services (less than 20) using 'docker stack deploy' CLI. Several internal overlay networks are created and some services are exposed to outside. Then multiple tests are run that basically access http endpoints exposed by one of services. After several hours dockerd on one of nodes hits the limit of open files.
This issue started to appear once we started to use 17.12.1-ce in our CI. Previously we used 17.06.2-ce for more than half a year and 'too many open files' was not observed there. 

**Steps to reproduce the issue:**
As in description

**Describe the results you received:**
- /var/log/messages gets flooded with "too many open files" error making systemd-journal constantly consume 100% of one CPU core. 
- docker CLI hangs

**Describe the results you expected:**
Limit of file descriptors is not hit and docker CLI works.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:	17.12.1-ce
 API version:	1.35
 Go version:	go1.9.4
 Git commit:	7390fc6
 Built:	Tue Feb 27 22:15:20 2018
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.1-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.4
  Git commit:	7390fc6
  Built:	Tue Feb 27 22:17:54 2018
  OS/Arch:	linux/amd64
  Experimental:	false
```

**Output of `docker info`:**
If this issue is encountered no CLI commands work, Pasting output from another execution which did not hit this issue yet:
```
Containers: 12
 Running: 8
 Paused: 0
 Stopped: 4
Images: 13
Server Version: 17.12.1-ce
Storage Driver: devicemapper
 Pool Name: vg00-docker--pool
 Pool Blocksize: 524.3kB
 Base Device Size: 10.74GB
 Backing Filesystem: xfs
 Udev Sync Supported: true
 Data Space Used: 7.73GB
 Data Space Total: 21.47GB
 Data Space Available: 13.75GB
 Metadata Space Used: 1.704MB
 Metadata Space Total: 218.1MB
 Metadata Space Available: 216.4MB
 Thin Pool Minimum Free Space: 2.147GB
 Deferred Removal Enabled: true
 Deferred Deletion Enabled: true
 Deferred Deleted Device Count: 0
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: k6zb3q7z4gz6v1qingmhlpg6f
 Is Manager: true
 ClusterID: n5y6hnmseoh9wxskc2237fbx6
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 2
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 10.9.4.157
 Manager Addresses:
  10.9.4.121:2377
  10.9.4.157:2377
  10.9.4.177:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 9b55aab90508bd389d7654c4baf173a981477d55
runc version: 9f9c96235cc97674e935002fc3d78361b696a69e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 3
Total Memory: 5.671GiB
Name: node-10-9-4-157
ID: KL7C:T2D7:OO2D:UKII:4NDF:KM7X:IEMK:WOCF:NTRY:XT44:6554:ADUR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 168
 Goroutines: 337
 System Time: 2018-05-14T19:27:01.265600845Z
 EventsListeners: 9
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 docker-registry:5000
 127.0.0.0/8
Registry Mirrors:
 http://docker-registry:5001/
Live Restore Enabled: false

WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
3 VMs with CentOS 7.2 on KVM.

First 1000 lines of warnings and errors from dockerd
[messages.log.filtered.1000.log](https://github.com/moby/moby/files/2002281/messages.log.filtered.1000.log)

```
[root@xxx]# cat lsof.docker.sock
(...)
dockerd 6368 root *526u  unix 0xffff880148d21a40      0t0  2724573 /var/run/docker.sock
dockerd 6368 root *527u  unix 0xffff880147495a40      0t0  2722306 /var/run/docker.sock
dockerd 6368 root *528u  unix 0xffff880148d003c0      0t0  2720591 /var/run/docker.sock
dockerd 6368 root *529u  unix 0xffff880147494000      0t0  2720592 /var/run/docker.sock
dockerd 6368 root *530u  unix 0xffff880148d030c0      0t0  2720593 /var/run/docker.sock
dockerd 6368 root *531u  unix 0xffff880148d22940      0t0  2747419 /var/run/docker.sock
dockerd 6368 root *532u  unix 0xffff880148d2bc00      0t0  2720595 /var/run/docker.sock
dockerd 6368 root *533u  unix 0xffff880147494b40      0t0  2722307 /var/run/docker.sock
dockerd 6368 root *534u  unix 0xffff880148d230c0      0t0  2747420 /var/run/docker.sock
dockerd 6368 root *535u  unix 0xffff880148d20780      0t0  2770958 /var/run/docker.sock
[root@xxx]# cat lsof.docker.sock | wc -l
65373
[root@xxx]# cat lsof.docker.sock | grep '/var/run/docker.sock' | wc -l
65372
```

**Comment by cpuguy83** [comment]
Please make sure you are on an to date version of docker (18.03.1 is the latest stable release).
Also are you sure these are not client's accessing the docker socket?

**Comment by pstrzelczak** [comment]
@cpuguy83 I performed a bit more detailed analysis of not closed file descriptors pointing to /var/run/docker.sock in dockerd process. I checked that contents of 'peer' field in kernel's unix_sock struct is reset meaning the client already closed the connection whereas dockerd keeps the descriptor not closed.

Also, I found (using strace) that dockerd keeps retrying accept calls:
```
(...)
[pid 14665]      0.000032 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000040 write(2, "time=\"2018-05-15T14:16:23.332229707Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
[pid 14665]      0.000029 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000038 write(2, "time=\"2018-05-15T14:16:23.332297269Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
[pid 14665]      0.000031 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000037 write(2, "time=\"2018-05-15T14:16:23.332365099Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
[pid 14665]      0.000029 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000038 write(2, "time=\"2018-05-15T14:16:23.332432282Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
[pid 14665]      0.000029 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000038 write(2, "time=\"2018-05-15T14:16:23.332498661Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
[pid 14665]      0.000030 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000038 write(2, "time=\"2018-05-15T14:16:23.332566849Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
[pid 14665]      0.000031 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000037 write(2, "time=\"2018-05-15T14:16:23.332635211Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
[pid 14665]      0.000032 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000042 write(2, "time=\"2018-05-15T14:16:23.332708288Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
[pid 14665]      0.000032 accept4(35, 0xc43a8f1ca8, 0xc43a8f1c9c, SOCK_CLOEXEC|SOCK_NONBLOCK) = -1 EMFILE (Too many open files)
[pid 14665]      0.000046 write(2, "time=\"2018-05-15T14:16:23.332779821Z\" level=warning msg=\"memberlist: Error accepting TCP connection: accept tcp [::]:7946: accept4: too many open files\"\n", 153) = 153
(...)
```

**Comment by pstrzelczak** [comment]
I haven't had a chance to play with 18.03.1 yet but are there any known bugs that lead to leaked file descriptors pointing to /var/run/docker.sock fixed there?

**Comment by cpuguy83** [comment]
dockerd is listening on docker.sock, any open connections to dockerd would show up as such.
The error you see there from strace is for overlay network syncing across the cluster and are not related to docker.sock specifically other than of course there are too many open files.

What is in `/proc/$(pidof dockerd)/fd`?

**Comment by pstrzelczak** [comment]
I know that the phenomenon visible in strace is a different story but I included it here as an interesting side effect because it seems 'overlay network syncing across the cluster' looks to eagerly repeat accept calls.

/proc/$(pidof dockerd)/fd:
[dockerd_fds.txt](https://github.com/moby/moby/files/2005505/dockerd_fds.txt)

**Comment by cpuguy83** [comment]
This would seem to indicate there are many open client connections.

**Comment by pstrzelczak** [comment]
Not really, because as said before it is very common in my case (see lsof based proof) that *only* dockerd keeps open file descriptors to /var/run/docker.sock. Of course, they most likely originate from past client connections that were already closed by client but apparently not closed by dockerd. I suspect that due to significant load on hypervisor where my VMs are deployed handling of docker API requests enters some uncommon path in either docker daemon or kernel code which finally leads to leaking file descriptors.

**Comment by pstrzelczak** [comment]
One thing that may help to find a root cause of this problem is that we have metricbeat configured on each host to collect docker stats every 10s. 
```
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.773885795Z" level=debug msg="Calling GET /containers/json?"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.781042704Z" level=debug msg="Calling GET /containers/4b11c06093e0ae2886368750d45a6bb570f4d612193da5690e392ddd2d622279/stats?stream=false"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.781268050Z" level=debug msg="Calling GET /containers/abfd929ba0a341dd367b8176b43f37de0fed6a9a613590a764818c3ecea41cc5/stats?stream=false"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.781297177Z" level=debug msg="Calling GET /containers/4277f755304956d78250447c979f005107027541051970330594a0112db5ad23/stats?stream=false"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.781498914Z" level=debug msg="Calling GET /containers/a81b3d83defb02061385c3dc2160f969d7fa63c8036ee963509456fc9fcf2e44/stats?stream=false"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.781509944Z" level=debug msg="Calling GET /containers/9d46555965d957d1687c831bb1ed0ea49e4b3cd4bd8ec6a4fe4ac66eb17d8d1b/stats?stream=false"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.781667473Z" level=debug msg="Calling GET /containers/6856c27100398f70fd2857761872b63965292276833470eb162077b9ac12ae9c/stats?stream=false"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.781680697Z" level=debug msg="Calling GET /containers/30ca1c03e71a59dad5aa4ad61c57e6ab13edc9acb8b87e7ed2599453fd91dad4/stats?stream=false"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.781850864Z" level=debug msg="Calling GET /containers/7a40b3ed1dbdadfd0b146bee63062e81887d053e7b5f5a86f8bb5a748e9c60bf/stats?stream=false"
May 11 12:23:22 localhost dockerd: time="2018-05-11T12:23:22.792393211Z" level=debug msg="Calling GET /containers/json?"
```

**Comment by cpuguy83** [comment]
To be clear, the mentioned patch should be the thing that fixes the FD leak.

**Comment by przemyslawstrzelczak9ld** [comment]
@cpuguy83 I just verified that the change in metricbeat has not been yet released in v6.2.4 (newest now) that we use.

Even if your statement about metricbeat code is correct, shouldn't I see same huge number of leaked descriptors in metricbeat process? As said before, I already caught moment /var/run/docker.sock is opened only by dockerd which hit 'too many open files' error.

```
[root@node-10-9-4-201 ~]# ls -al /proc/`pidof metricbeat`/fd/
total 0
dr-x------ 2 root root  0 May 15 12:06 .
dr-xr-xr-x 9 root root  0 May 15 04:22 ..
lr-x------ 1 root root 64 May 15 12:06 0 -> /dev/null
lrwx------ 1 root root 64 May 15 12:06 1 -> socket:[169366570]
lrwx------ 1 root root 64 May 15 12:06 10 -> socket:[170101345]
lrwx------ 1 root root 64 May 15 12:06 11 -> socket:[170101788]
lrwx------ 1 root root 64 May 15 12:06 13 -> socket:[170101346]
lrwx------ 1 root root 64 May 15 12:06 2 -> socket:[169366570]
lr-x------ 1 root root 64 May 15 12:06 3 -> /dev/urandom
lrwx------ 1 root root 64 May 15 12:06 4 -> anon_inode:[eventpoll]
l-wx------ 1 root root 64 May 15 12:06 5 -> /var/log/metricbeat/metricbeat
lrwx------ 1 root root 64 May 15 12:06 6 -> socket:[170101341]
lrwx------ 1 root root 64 May 15 12:06 7 -> socket:[170101342]
lrwx------ 1 root root 64 May 15 12:06 8 -> socket:[170101343]
lrwx------ 1 root root 64 May 15 12:06 9 -> socket:[170101344]
```

**Comment by pstrzelczak** [comment]
@cpuguy83 I just verified that the change in metricbeat has not been yet released in v6.2.4 (newest now) that we use.

Even if your statement about metricbeat code is correct, shouldn't I see same huge number of leaked descriptors in metricbeat process? As said before, I already caught moment /var/run/docker.sock is opened only by dockerd which hit 'too many open files' error.

```
[root@node-10-9-4-201 ~]# ls -al /proc/`pidof metricbeat`/fd/
total 0
dr-x------ 2 root root  0 May 15 12:06 .
dr-xr-xr-x 9 root root  0 May 15 04:22 ..
lr-x------ 1 root root 64 May 15 12:06 0 -> /dev/null
lrwx------ 1 root root 64 May 15 12:06 1 -> socket:[169366570]
lrwx------ 1 root root 64 May 15 12:06 10 -> socket:[170101345]
lrwx------ 1 root root 64 May 15 12:06 11 -> socket:[170101788]
lrwx------ 1 root root 64 May 15 12:06 13 -> socket:[170101346]
lrwx------ 1 root root 64 May 15 12:06 2 -> socket:[169366570]
lr-x------ 1 root root 64 May 15 12:06 3 -> /dev/urandom
lrwx------ 1 root root 64 May 15 12:06 4 -> anon_inode:[eventpoll]
l-wx------ 1 root root 64 May 15 12:06 5 -> /var/log/metricbeat/metricbeat
lrwx------ 1 root root 64 May 15 12:06 6 -> socket:[170101341]
lrwx------ 1 root root 64 May 15 12:06 7 -> socket:[170101342]
lrwx------ 1 root root 64 May 15 12:06 8 -> socket:[170101343]
lrwx------ 1 root root 64 May 15 12:06 9 -> socket:[170101344]
```

**Comment by pstrzelczak** [comment]
I might be soon asked to delete the environment with the instance of this problem. Could you recommend what kind of logs I could collect before?

**Comment by cpuguy83** [comment]
The thing is, in dockerd the client connection is handled by the stdlib
http server. These are automatically closed out except in a few specific
cases (where the daemon hijacks the connection to enable 2-way
communication for container attach). The old metricsbeat code was
*definitely* incorrect (the docker client was waiting on a nil channel,
which will always block because metricsbeat wasn't setting the channel, in
order to close out the response body).

Maybe we can get a goroutine stack dump from docker to see what might be
still running. You can get this from `curl --unix-socket
/var/run/docker.sock http://./debug/pprof/goroutine\?debug\=2`

**Comment by pstrzelczak** [comment]
@cpuguy83 is there any other way to get a goroutine stack dump because due to fd limit already hit in dockerd above curl connection is not accepted :/ ?

**Comment by cpuguy83** [comment]
Oh damn... :(
You could send SIGQUIT which will dump the stack to stderr (so it will be
in the daemon logs). But this will also trigger the daemon to exit.
I don't think there is other troubleshooting that could be done anyway, so
probably best to just go ahead and do that.

**Comment by cyrankiewicz** [comment]
:+1:

**Comment by pstrzelczak** [comment]
@cpuguy83 we checked number of open files in dockerd in other environments in our lab and it seems many of them are affected by this problem. Luckily, some of them did not hit 64K open files limit so we managed to dump goroutines of dockerd on one affected environment. Basically this looks like a deadlock:
```
goroutine profile: total 65143
64415 @ 0x11a34fc 0x11a35ee 0x11b4c24 0x11b493d 0x11db89e 0x2069a4d 0x24e9bbf 0x20bb224 0x20c23e9 0x1b0f97a 0x1b0e078 0x1b0e4d3 0x16d510b 0x1b0d77e 0x1b22b86 0x1456f84 0x16d6a5b 0x1b21340 0x1459c64 0x1455e5d 0x11d2de1
#       0x11b493c       sync.runtime_SemacquireMutex+0x3c                                                                                                                               /usr/local/go/src/runtime/sema.go:71
#       0x11db89d       sync.(*Mutex).Lock+0xed                                                                                                                                         /usr/local/go/src/sync/mutex.go:134
#       0x2069a4c       github.com/docker/docker/container.(*State).IsRunning+0x2c                                                                                                      /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/container/state.go:250
#       0x24e9bbe       github.com/docker/docker/daemon.(*Daemon).ContainerStats+0xbe                                                                                                   /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/stats.go:31
#       0x20bb223       github.com/docker/docker/api/server/router/container.(*containerRouter).getContainersStats+0x1e3                                                                /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/api/server/router/container/container_routes.go:76
#       0x20c23e8       github.com/docker/docker/api/server/router/container.(*containerRouter).(github.com/docker/docker/api/server/router/container.getContainersStats)-fm+0x68       /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/api/server/router/container/container.go:56
#       0x1b0f979       github.com/docker/docker/api/server/router.cancellableHandler.func1+0xd9                                                                                        /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/api/server/router/local.go:92
#       0x1b0e077       github.com/docker/docker/api/server/middleware.ExperimentalMiddleware.WrapHandler.func1+0xd7                                                                    /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/api/server/middleware/experimental.go:27
#       0x1b0e4d2       github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1+0x422                                                                        /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/api/server/middleware/version.go:62
#       0x16d510a       github.com/docker/docker/pkg/authorization.(*Middleware).WrapHandler.func1+0x83a                                                                                /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/pkg/authorization/middleware.go:59
#       0x1b0d77d       github.com/docker/docker/api/server/middleware.DebugRequestMiddleware.func1+0x17d                                                                               /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/api/server/middleware/debug.go:22
#       0x1b22b85       github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1+0x1c5                                                                                       /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/api/server/server.go:137
#       0x1456f83       net/http.HandlerFunc.ServeHTTP+0x43                                                                                                                             /usr/local/go/src/net/http/server.go:1918
#       0x16d6a5a       github.com/docker/docker/vendor/github.com/gorilla/mux.(*Router).ServeHTTP+0x22a                                                                                /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/gorilla/mux/mux.go:103
#       0x1b2133f       github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP+0x6f                                                                                             /root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/api/server/router_swapper.go:29
#       0x1459c63       net/http.serverHandler.ServeHTTP+0xb3                                                                                                                           /usr/local/go/src/net/http/server.go:2619
#       0x1455e5c       net/http.(*conn).serve+0x71c                                                                                                                                    /usr/local/go/src/net/http/server.go:1801

```

Unfortunately it seems that detailed dump (debug=2) is limited to 64MB so not information is available.
Attaching debug=1 and debug=2 dumps
[goroutine_debug.tar.gz](https://github.com/moby/moby/files/2015945/goroutine_debug.tar.gz)

**Comment by pstrzelczak** [comment]
We created a simple script that traverses all environment in our LAB and we have the following observation
- *most* of environments with 17.12-1 older more than 1 day hit this problem
- *none* of environments with 17.06-2 (including ones older that 1 day) hit this problem.

As a result I am starting to think that it is a problem in 17.12-1 (metricbeat version was basically the same).

**Comment by cpuguy83** [comment]
This is blocked here, waiting on a response from containerd:

```

goroutine 16003775 [select, 1341 minutes]:
github.com/docker/docker/vendor/google.golang.org/grpc/transport.(*Stream).Header(0xc4600250e0, 0x2aeb618, 0xc422b93030, 0x38c2a40)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/google.golang.org/grpc/transport/transport.go:256 +0x146
github.com/docker/docker/vendor/google.golang.org/grpc.recvResponse(0x7fedaeaa18a0, 0xc45ad58660, 0xc4200e0ae0, 0xc4200e0b00, 0x38b9200, 0x3939c30, 0x0, 0x0, 0x0, 0x0, ...)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/google.golang.org/grpc/call.go:64 +0xac
github.com/docker/docker/vendor/google.golang.org/grpc.invoke(0x7fedaeaa18a0, 0xc45ad58660, 0x2a916ad, 0x27, 0x28fad80, 0xc4355fffa0, 0x28fae60, 0xc45cc22160, 0xc42007a680, 0x0, ...)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/google.golang.org/grpc/call.go:279 +0xcfe
github.com/docker/docker/vendor/github.com/containerd/containerd.namespaceInterceptor.unary(0x2a4d539, 0x4, 0x7fedaea95000, 0xc420014018, 0x2a916ad, 0x27, 0x28fad80, 0xc4355fffa0, 0x28fae60, 0xc45cc22160, ...)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/containerd/containerd/grpc.go:18 +0xf4
github.com/docker/docker/vendor/github.com/containerd/containerd.(namespaceInterceptor).(github.com/docker/docker/vendor/github.com/containerd/containerd.unary)-fm(0x7fedaea95000, 0xc420014018, 0x2a916ad, 0x27, 0x28fad80, 0xc4355fffa0, 0x28fae60, 0xc45cc22160, 0xc42007a680, 0x2aeb5f0, ...)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/containerd/containerd/grpc.go:34 +0xf4
github.com/docker/docker/vendor/google.golang.org/grpc.Invoke(0x7fedaea95000, 0xc420014018, 0x2a916ad, 0x27, 0x28fad80, 0xc4355fffa0, 0x28fae60, 0xc45cc22160, 0xc42007a680, 0x0, ...)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/google.golang.org/grpc/call.go:141 +0xdd
github.com/docker/docker/vendor/github.com/containerd/containerd/api/services/tasks/v1.(*tasksClient).Get(0xc45cc22150, 0x7fedaea95000, 0xc420014018, 0xc4355fffa0, 0x0, 0x0, 0x0, 0xc422b936e8, 0x13c8c97, 0x28a80e0)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/containerd/containerd/api/services/tasks/v1/tasks.pb.go:450 +0xd2
github.com/docker/docker/vendor/github.com/containerd/containerd.(*process).Status(0xc4684965d0, 0x38b9a00, 0xc420014018, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc4355fff80, ...)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/containerd/containerd/process.go:201 +0x16d
github.com/docker/docker/vendor/github.com/containerd/containerd.(*process).Delete(0xc4684965d0, 0x38b9a00, 0xc420014018, 0x0, 0x0, 0x0, 0xc45b93efc0, 0x38c3040, 0xc4684965d0)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/vendor/github.com/containerd/containerd/process.go:178 +0x131
github.com/docker/docker/libcontainerd.(*client).Exec(0xc4202fc3f0, 0x38b9a40, 0xc45b93e2a0, 0xc44b4e2f40, 0x40, 0xc44f04f200, 0x40, 0xc47373c340, 0x0, 0xc45e2794d0, ...)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/libcontainerd/client_daemon.go:362 +0x5fa
github.com/docker/docker/daemon.(*Daemon).ContainerExecStart(0xc420238000, 0x7fedaeadb0f8, 0xc45b93e2a0, 0xc44f04f200, 0x40, 0x0, 0x0, 0x38985c0, 0xc464309900, 0x38985c0, ...)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/exec.go:245 +0x9ca
github.com/docker/docker/daemon.(*cmdProbe).run(0xc423f94239, 0x7fedaeadb0f8, 0xc45b93e2a0, 0xc420238000, 0xc4642538c0, 0x0, 0x0, 0x11cceb0)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/health.go:95 +0x5b8
github.com/docker/docker/daemon.monitor.func1(0x3898540, 0xc423f94239, 0x7fedaeadb0f8, 0xc45b93e2a0, 0xc420238000, 0xc4642538c0, 0xc45b93e300, 0xbeb766a0317526fe, 0xd561a93f8347, 0x3916b60)
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/health.go:200 +0xb0
created by github.com/docker/docker/daemon.monitor
	/root/rpmbuild/BUILD/src/engine/.gopath/src/github.com/docker/docker/daemon/health.go:198 +0x336
```

Which does not seem like it will come.
There are a number of issues fixes in containerd which are included in 18.03.1 that should solve this.

**Comment by pstrzelczak** [comment]
Shall we label this bug with 17.12.1 and mark as a duplicate of libcontainerd issue (could you please point one)?

**Comment by cpuguy83** [comment]
It could be a couple of things, there are some fixes in containerd 1.0.2 (https://github.com/containerd/containerd/releases/tag/v1.0.2), and also down in runc (specifically https://github.com/opencontainers/runc/pull/1698).

**Comment by hartemiev-vs** [comment]
I can confirm that this issue affects 17.12.1 version of Docker. 

On one of release notes I found this PR https://github.com/elastic/beats/pull/11294

Does any body uses Metricbeat 7.0.0 with 17.12.1 version of Docker? Was it really fixed there? 

We use Metricbeat 6.7.2 where it is reproducible.

**Comment by tiredpixel** [comment]
We appear to have just hit this, or at least something similar to this, on Docker 18.09.7, Metricbeat 7.2.0, also using the Metricbeat attaching to the Docker socket every `10 s` (the default). We've been running this Docker and Metricbeat setup for months, but only enabled the Docker monitoring last week.

Metricbeat error:
```
Jul 12 07:45:01 XXX metricbeat[19143]: 2019-07-12T07:45:01.724Z        ERROR        [docker]        event/event.go:99        Error watching for docker events: %verror d
uring connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.22/events?since=1562917415: dial unix /var/run/docker.sock: socket: too many open files
Jul 12 07:45:02 XXX metricbeat[19143]: 2019-07-12T07:45:02.725Z        ERROR        [docker]        event/event.go:99        Error watching for docker events: %verror d
uring connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.22/events?since=1562917415: dial unix /var/run/docker.sock: socket: too many open files
Jul 12 07:45:03 XXX metricbeat[19143]: 2019-07-12T07:45:03.726Z        ERROR        [docker]        event/event.go:99        Error watching for docker events: %verror d
uring connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.22/events?since=1562917415: dial unix /var/run/docker.sock: socket: too many open files
```

Dist:
```
CentOS Linux release 7.6.1810 (Core)
```

Docker:
```
Client:
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        2d0083d
 Built:             Thu Jun 27 17:56:06 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.7
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.8
  Git commit:       2d0083d
  Built:            Thu Jun 27 17:26:28 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```

Runc:
```
runc version 1.0.0-rc8
commit: 425e105d5a03fabd737a126ad93d62a9eeede87f
spec: 1.0.1-dev
```

Containerd:
```
containerd containerd.io 1.2.6 894b81a4b802e4eb2a91d1ce216b8817763c29fb
```

Metricbeat (CentOS package; this is the one connecting to the Docker socket):
```
metricbeat version 7.2.0 (amd64), libbeat 7.2.0 [9ba65d864ca37cd32c25b980dbb4020975288fc0 built 2019-06-20 15:07:31 +0000 UTC]
```

Metricbeat (Docker container; we're running both, doing different things; this doesn't connect to the Docker socket, but I include it anyway, for clarity):
```
docker.elastic.co/beats/metricbeat:7.2.0@sha256:308cb2581c762d2517d39009cec106e8883162f84206b1dc8860a5df30e5e3ba
```

```
# lsof /var/run/docker.sock | wc -l
1025
# ls /proc/31173/fd | wc -l  # dockerd pid
1158
```

```
# curl --unix-socket /var/run/docker.sock http://./debug/pprof/goroutine\?debug\=2 > go.debug
```
[go.log](https://github.com/moby/moby/files/3385807/go.log)

**Comment by tiredpixel** [comment]
This has just happened to a cluster of hours again. I'm disabling the use of that Metricbeat module, as otherwise it introduces an instability causing a cluster crash every couple of days.

**Comment by liozzazhang** [comment]
> We appear to have just hit this, or at least something similar to this, on Docker 18.09.7, Metricbeat 7.2.0, also using the Metricbeat attaching to the Docker socket every `10 s` (the default). We've been running this Docker and Metricbeat setup for months, but only enabled the Docker monitoring last week.
> 
> Metricbeat error:
> 
> ```
> Jul 12 07:45:01 XXX metricbeat[19143]: 2019-07-12T07:45:01.724Z        ERROR        [docker]        event/event.go:99        Error watching for docker events: %verror d
> uring connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.22/events?since=1562917415: dial unix /var/run/docker.sock: socket: too many open files
> Jul 12 07:45:02 XXX metricbeat[19143]: 2019-07-12T07:45:02.725Z        ERROR        [docker]        event/event.go:99        Error watching for docker events: %verror d
> uring connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.22/events?since=1562917415: dial unix /var/run/docker.sock: socket: too many open files
> Jul 12 07:45:03 XXX metricbeat[19143]: 2019-07-12T07:45:03.726Z        ERROR        [docker]        event/event.go:99        Error watching for docker events: %verror d
> uring connect: Get http://%2Fvar%2Frun%2Fdocker.sock/v1.22/events?since=1562917415: dial unix /var/run/docker.sock: socket: too many open files
> ```
> 
> Dist:
> 
> ```
> CentOS Linux release 7.6.1810 (Core)
> ```
> 
> Docker:
> 
> ```
> Client:
>  Version:           18.09.7
>  API version:       1.39
>  Go version:        go1.10.8
>  Git commit:        2d0083d
>  Built:             Thu Jun 27 17:56:06 2019
>  OS/Arch:           linux/amd64
>  Experimental:      false
> 
> Server: Docker Engine - Community
>  Engine:
>   Version:          18.09.7
>   API version:      1.39 (minimum version 1.12)
>   Go version:       go1.10.8
>   Git commit:       2d0083d
>   Built:            Thu Jun 27 17:26:28 2019
>   OS/Arch:          linux/amd64
>   Experimental:     false
> ```
> 
> Runc:
> 
> ```
> runc version 1.0.0-rc8
> commit: 425e105d5a03fabd737a126ad93d62a9eeede87f
> spec: 1.0.1-dev
> ```
> 
> Containerd:
> 
> ```
> containerd containerd.io 1.2.6 894b81a4b802e4eb2a91d1ce216b8817763c29fb
> ```
> 
> Metricbeat (CentOS package; this is the one connecting to the Docker socket):
> 
> ```
> metricbeat version 7.2.0 (amd64), libbeat 7.2.0 [9ba65d864ca37cd32c25b980dbb4020975288fc0 built 2019-06-20 15:07:31 +0000 UTC]
> ```
> 
> Metricbeat (Docker container; we're running both, doing different things; this doesn't connect to the Docker socket, but I include it anyway, for clarity):
> 
> ```
> docker.elastic.co/beats/metricbeat:7.2.0@sha256:308cb2581c762d2517d39009cec106e8883162f84206b1dc8860a5df30e5e3ba
> ```
> 
> ```
> # lsof /var/run/docker.sock | wc -l
> 1025
> # ls /proc/31173/fd | wc -l  # dockerd pid
> 1158
> ```
> 
> ```
> # curl --unix-socket /var/run/docker.sock http://./debug/pprof/goroutine\?debug\=2 > go.debug
> ```
> 
> [go.log](https://github.com/moby/moby/files/3385807/go.log)

I think you just hit the limitation of 1024

**Comment by thaJeztah** [comment]
Closing as this went stale
